sway-ir/src/optimize/constants.rs

```diff
@@ -22,12 +23,85 @@ pub fn combine_constants(context: &mut Context, function: &Function) -> Result<b
             continue;
         }
 
+        if fold_cmp(context, function) {
+            modified = true;
+            continue;
+        }
+
+        if fold_cbr(context, function)? {
+            modified = true;
+            continue;
+        }
+
         // Other passes here... always continue to the top if pass returns true.
         break;
     }
+
     Ok(modified)
 }
 
+fn fold_cbr(context: &mut Context, function: &Function) -> Result<bool, IrError> {
+    let candidate = function
+        .instruction_iter(context)
+        .find_map(|(_, inst_val)| match &context.values[inst_val.0].value {
+            ValueDatum::Instruction(Instruction::ConditionalBranch {
+                cond_value,
+                true_block,
+                false_block,
+            }) if cond_value.is_constant(context) => {
+                match cond_value.get_constant(context).unwrap().value {
+                    ConstantValue::Bool(true) => Some(Ok((inst_val, true_block))),
+                    ConstantValue::Bool(false) => Some(Ok((inst_val, false_block))),
+                    _ => Some(Err(IrError::VerifyConditionExprNotABool)),
+                }
+            }
+            _ => None,
+        });
+
+    match candidate {
+        Some(res) => {
+            let (cbr, dest) = res?;
+            context.values[cbr.0].value = ValueDatum::Instruction(Instruction::Branch(*dest));
+            Ok(true)
+        }
+        None => Ok(false),
```

Comment:
You can use `Option::transpose()` here to get a `Result<Option<...>>` and use `?` on line 59.  Then the `candidate` will be an `Option` which you can `.map_or()` into a `Result<bool>`.  So the match expression could be replaced with:
```
Ok(candidate.map_or(false, |(cbr, dest)| { ... ; true }))
```

Do we not have a direct `replace_instruction()` which does the gnarly `context.values[...].value = ...` line?  I reckon it should be a method on `Value` to replace itself with an alternative.
```
fn replace(&self, context: &mut Context, other: Value) { ... }
```